README.md
# jhubcode README

This Extention allows you to edit files on the UVic JupyterLab server from VSCode

## Usage

Press `cmd-shift-P` to run the `Init JupyterLab Code` then enter your JupyterLab API token (which can be gotten at [https://jhub.csc.uvic.ca/hub/token]()) and your JupyterLab username.

Reload the files with the button in the lower status bar.

Open a new terminal by running the command `Start Jupyterlab Terminal`.

## Known Issues

Cannot delete folders
Creating a new file works but shows an error

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of JHubCode

CHANGELOG.md
# Change Log

All notable changes to the "jhubcode" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.js
const { JHubFS } = require('./fileSystemProvider');
const { TerminalViewProvider } = require('./terminalViewProvider');

const vscode = require('vscode');

function activate(context) {
	const FS = new JHubFS();
    let initialized = false;

	context.subscriptions.push(vscode.workspace.registerFileSystemProvider('jhubfs', FS, { isCaseSensitive: true }));

	context.subscriptions.push(vscode.commands.registerCommand('jhubcode.init', async () => {
        if (initialized) {
            return;
        }

		if (!FS.token) {
			FS.token = await vscode.window.showInputBox({ prompt: 'Api Token' });
		}

		if (!FS.username) {
			FS.username = await vscode.window.showInputBox({ prompt: 'Username' });
		}
		console.log('init')
		try {
			initialized = await FS.init();
		} catch (e) {
			console.log('error');

			console.log(e);
		}

		if (initialized) {
            vscode.workspace.updateWorkspaceFolders(0, 0, { uri: vscode.Uri.parse('jhubfs:/'), name: "Jupyterlab" });

			FS.loadFiles();

			context.subscriptions.push(vscode.commands.registerCommand('jhubcode.load', async () => {
				if (initialized) {
					FS.loadFiles();
				}
			}));
		
			const reloadIcon = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 100);
			reloadIcon.command = 'jhubcode.load';
			context.subscriptions.push(reloadIcon);
			reloadIcon.text = 'reload jupyterlab files'
			reloadIcon.show();
		}
    }));

	context.subscriptions.push(
		vscode.commands.registerCommand('jhubcode.terminalStart', () => {
			if (initialized) {
				new TerminalViewProvider(context.extensionUri, FS);
			} else {
				vscode.window.showInformationMessage('Error: JHubCode not initialized');
			}
		})
	);
}

function deactivate() {}

module.exports = {
	activate,
	deactivate
}


src/terminalViewProvider.js
const vscode = require('vscode');
var https = require('https');
var WebSocketClient = require('websocket').client;

class TerminalViewProvider {
	constructor(extensionUri, FS) {
		const column = vscode.window.activeTextEditor
			? vscode.window.activeTextEditor.viewColumn
			: undefined;

		this._panel = vscode.window.createWebviewPanel(
			TerminalViewProvider.viewType,
			'JupyterLab Terminal',
			column || vscode.ViewColumn.One,
			getWebviewOptions(extensionUri),
		);

		this._extensionUri = extensionUri;
		this._FS = FS;
		this.connected = false;
		this.termID;

		this._update();

		this._panel.onDidDispose(() => this.dispose(), null, this._disposables);

		this._panel.onDidChangeViewState(
			e => {
				if (this._panel.visible) {
					this._update();
				}
			},
			null,
			this._disposables
		);
	}

	doRefactor() {
		this._panel.webview.postMessage({ command: 'refactor' });
	}

	dispose() {
		this._panel.dispose();
		var req = https.request({
			hostname: "jhub.csc.uvic.ca",
			port: 443,
			path: `/user/${this._FS.username}/api/terminals/${this.termID}?token=${this._FS.token}`,
			method: 'DELETE',
		  }, () => {})
		req.end();
	}

	_update() {
		const panel = this._panel;
		panel.title = 'Jupyterlab Terminal';
		if (panel.webview) {
			panel.webview.html = this._getHtmlForWebview(panel.webview);
			if (!this.connected) {
				this.connect();
			}
		}

	}

	connect() {
		const panel = this._panel;
		const FS = this._FS;

		var req = https.request({
			hostname: "jhub.csc.uvic.ca",
			port: 443,
			path: `/user/${FS.username}/api/terminals?token=${FS.token}`,
			method: 'POST',
		  }, (res) => {
			res.setEncoding('utf8');

			var body = '';

			res.on('data', (chunk) => {
				body = body + chunk;
			});

			res.on('end', () => {
				if (res.statusCode == 200) {
					var client = new WebSocketClient();
	
					client.on('connectFailed', (error) => {
						panel.webview.postMessage(['stdin', 'Connect Error: ' + error.toString()]);
						this.connected = false;
					});
			
					client.on('connect', (connection) => {
						panel.webview.postMessage(['stdin', 'WebSocket Client Connected']);
						this.connected = true;
						connection.on('message', function(message) {
							if (message.type === 'utf8') {
								panel.webview.postMessage(JSON.parse(message.utf8Data));
							}
						});
		
						panel.webview.onDidReceiveMessage(
							message => {
									connection.sendUTF(JSON.stringify(message));
							},
							null,
							this._disposables
						);
	
						connection.on('error', (error)  => {
							panel.webview.postMessage(['stdin', "Connection Error: " + error.toString()]);
							this.connected = false;
						});
						connection.on('close', () => {
							panel.webview.postMessage(['stdin', 'echo-protocol Connection Closed']);
							this.connected = false;
						});
					});

					this.termID = JSON.parse(body).name;
			
					client.connect(`wss://jhub.csc.uvic.ca/user/${FS.username}/terminals/websocket/${this.termID}?token=${FS.token}`);
				}
			});
		});
		
		req.end();
	}

	_getHtmlForWebview(webview) {
		const scriptPathOnDisk = vscode.Uri.joinPath(this._extensionUri, 'src', 'terminal.js');
		const scriptUri = (scriptPathOnDisk).with({ 'scheme': 'vscode-resource' });

		const scriptXtermOnDisk = vscode.Uri.joinPath(this._extensionUri, 'node_modules', 'xterm', 'lib', 'xterm.js');
		const scriptXtermUri = (scriptXtermOnDisk).with({ 'scheme': 'vscode-resource' });

		const scriptXtermFitOnDisk = vscode.Uri.joinPath(this._extensionUri, 'node_modules', 'xterm-addon-fit', 'lib', 'xterm-addon-fit.js');
		const scriptXtermFitUri = (scriptXtermFitOnDisk).with({ 'scheme': 'vscode-resource' });

		const stylesPathXterm = vscode.Uri.joinPath(this._extensionUri, 'node_modules', 'xterm', 'css', 'xterm.css');
		const stylesXtermUri = webview.asWebviewUri(stylesPathXterm);

		const nonce = getNonce();

		return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<!--
					Use a content security policy to only allow loading images from https or from our extension directory,
					and only allow scripts that have a specific nonce.
				-->
				<title>Jupyterlab Terminal</title>
				<link href="${stylesXtermUri}" rel="stylesheet">
				<script nonce="${nonce}" src="${scriptXtermUri}"></script>
				<script nonce="${nonce}" src="${scriptXtermFitUri}"></script>
			</head>
			<body>
				<div id="terminal" style="width: 100vw; height: 100vh;"></div>
			</body>
			<script id="helper" nonce="${nonce}" src="${scriptUri}"></script>
			</html>`;
	}
}

function getNonce() {
	let text = '';
	const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < 32; i++) {
		text += possible.charAt(Math.floor(Math.random() * possible.length));
	}
	return text;
}

function getWebviewOptions(extensionUri) {
	return {
		enableScripts: true,
		localResourceRoots: [
			vscode.Uri.joinPath(extensionUri, 'src'),
			vscode.Uri.joinPath(extensionUri, 'node_modules', 'xterm', 'lib'),
			vscode.Uri.joinPath(extensionUri, 'node_modules', 'xterm', 'css'),
			vscode.Uri.joinPath(extensionUri, 'node_modules', 'xterm-addon-fit', 'lib')
		]
	};
}

module.exports = {
	TerminalViewProvider,
	getWebviewOptions
}

src/fileSystemProvider.js
const path = require('path');
const vscode = require('vscode');
var https = require('https');

const output = vscode.window.createOutputChannel("JHub");

const prefix = 'jhubfs'

const sendRequest = ({options, data}) => {
    return new Promise((resolve, reject) => {
        var req = https.request(options, function (res) {
            res.setEncoding('utf8');

            var body = '';

            res.on('data', function (chunk) {
                body = body + chunk;
            });

            res.on('end',function(){
                if (res.statusCode == 200) {
                    resolve(JSON.parse(body));
                } else {
                    reject();
                }
            });

        });
        
        req.on('error', function (e) {
            output.appendLine(e);
            reject();
        });

        if (data) {
            req.write(JSON.stringify(data));
        }
        req.end();
    });

}
class File {
    constructor(name) {
        this.type = vscode.FileType.File;
        this.ctime = Date.now();
        this.mtime = Date.now();
        this.size = 0;
        this.mimetype = "";
        this.name = name;
    }
}

class Directory {
    constructor(name) {
        this.type = vscode.FileType.Directory;
        this.ctime = Date.now();
        this.mtime = Date.now();
        this.size = 0;
        this.name = name;
        this.entries = new Map();
    }
}

class JHubFS {
    constructor() {
        this.root = new Directory('')

        this.username;
        this.token;

        this._emitter = new vscode.EventEmitter();
        this._bufferedEvents = [];
        this._fireSoonHandle;
    
        this.onDidChangeFile = this._emitter.event;

    }
    async init() {
        try {
            const res = await sendRequest({
                options:  {
                    hostname: "jhub.csc.uvic.ca",
                    port: 443,
                    path: `/user/${this.username}/api/status?token=${this.token}`,
                    method: 'GET',
                }
            })

            if (res.started) {
                return true;
            } else {
                vscode.window.showInformationMessage('Error: Server not started');
                return false;
            }

        } catch (e) {
            vscode.window.showInformationMessage('Error: Server not started');
            return false
        }

    }
    parseDir(dir, top=false) {
        const uri = `${prefix}:/${dir.path}/`;
        if (dir.type != 'directory') {
            throw vscode.FileSystemError.FileNotADirectory(uri);
        }
        if (!top) {
            this.createDirectory(vscode.Uri.parse(uri));
        }
        dir.content.forEach(item => {
            if (item.type == 'file') {
                const fileUri = `${prefix}:/${item.path}`;
                this.createFileEntry(vscode.Uri.parse(fileUri), item, { create: true, overwrite: true });
            } else if (item.type == 'directory') {
                this.loadPath(item.path).then(data => this.parseDir(data));   
            }
        });
        this._fireSoon({ type: vscode.FileChangeType.Changed, uri });
    }
    loadPath(path) {
        return sendRequest({
            options:  {
                hostname: "jhub.csc.uvic.ca",
                port: 443,
                path: `/user/${this.username}/api/contents${path.startsWith('/')?'':'/'}${path}?token=${this.token}`,
                method: 'GET',
              }
        });
    }

    loadFiles() {
        this.root = new Directory('')
        output.appendLine('loading files')

        this.loadPath('').then(data => {
            output.appendLine('got data')
            this.parseDir(data, true);
        });   
    }

    // --- manage file metadata

    stat(uri) {
        return this._lookup(uri, false);
    }

    readDirectory(uri) {
        const entry = this._lookupAsDirectory(uri, false);
        const result = [];
        for (const [name, child] of entry.entries) {
            result.push([name, child.type]);
        }
        return result;
    }

    // --- manage file contents

    readFile(uri) {
        const entry = this._lookupAsFile(uri, false);
        if (entry) {
            if (entry.data) {
                return entry.data;
            } else {
                return new Promise((resolve, reject) => {
                    this.loadPath(uri.path).then(file => {
                        let content = Buffer.from('');
                        switch (file.format) {
                            default:
                            case 'text':
                                content = Buffer.from(file.content);
                                break;
                            case 'base64':
                                content = Buffer.from(file.content, 'base64');
                                break;
                        }
                        resolve(content);
                    }).catch(err => {
                        reject(err);
                    });
                });
            }
        }
        throw vscode.FileSystemError.FileNotFound();
    }

    createFileEntry(uri, file, options) {
        const { create, overwrite } = options;

        const basename = path.posix.basename(uri.path);
        const parent = this._lookupParentDirectory(uri);
        let entry = parent.entries.get(basename);
        if (entry instanceof Directory) {
            throw vscode.FileSystemError.FileIsADirectory(uri);
        }
        if (!entry && !create) {
            throw vscode.FileSystemError.FileNotFound(uri);
        }
        if (entry && create && !overwrite) {
            throw vscode.FileSystemError.FileExists(uri);
        }
        if (!entry) {
            entry = new File(basename);
            parent.entries.set(basename, entry);
            this._fireSoon({ type: vscode.FileChangeType.Created, uri });
        }
        entry.ctime = Date.parse(file.created);
        entry.mtime = Date.parse(file.last_modified);
        entry.size = file.size;
        entry.data = undefined;

        this._fireSoon({ type: vscode.FileChangeType.Changed, uri });
    }

    writeFile(uri, content, options) {
        return sendRequest({
            options: {
                hostname: "jhub.csc.uvic.ca",
                port: 443,
                path: `/user/${this.username}/api/contents${uri.path}?token=${this.token}`,
                method: 'PUT',
            }, 
            data: {
                content: content.toString(),
                format: 'text',
                path:  uri.path.substring(1),
                type: 'file'
            }
        })
    }

    // --- manage files/folders

    rename(oldUri, newUri, options) {
        // TODO
        if (!options.overwrite && this._lookup(newUri, true)) {
            throw vscode.FileSystemError.FileExists(newUri);
        }

        sendRequest({
            options: {
                hostname: "jhub.csc.uvic.ca",
                port: 443,
                path: `/user/${this.username}/api/contents${oldUri.path}?token=${this.token}`,
                method: 'PATCH',
            }, data: {
                'path': newUri.path.substring(1)
            }
        }).then(() => {
            const entry = this._lookup(oldUri, false);
            const oldParent = this._lookupParentDirectory(oldUri);
    
            const newParent = this._lookupParentDirectory(newUri);
            const newName = path.posix.basename(newUri.path);
    
            oldParent.entries.delete(entry.name);
            entry.name = newName;
            newParent.entries.set(newName, entry);

            this._fireSoon(
                { type: vscode.FileChangeType.Deleted, uri: oldUri },
                { type: vscode.FileChangeType.Created, uri: newUri }
            );
        });
    }

    delete(uri) {
        const dirname = uri.with({ path: path.posix.dirname(uri.path) });
        const basename = path.posix.basename(uri.path);
        const parent = this._lookupAsDirectory(dirname, false);
        if (!parent.entries.has(basename)) {
            throw vscode.FileSystemError.FileNotFound(uri);
        }
        sendRequest({
            options: {
                hostname: "jhub.csc.uvic.ca",
                port: 443,
                path: `/user/${this.username}/api/contents${uri.path}?token=${this.token}`,
                method: 'DELETE',
            }, 
        }).then(() => {
            this._fireSoon({ type: vscode.FileChangeType.Changed, uri: dirname }, { uri, type: vscode.FileChangeType.Deleted });
        });
    }

    createDirectory(uri) {
        const basename = path.posix.basename(uri.path);
        const dirname = uri.with({ path: path.posix.dirname(uri.path) });
        const parent = this._lookupAsDirectory(dirname, false);
        
        const entry = new Directory(basename);
        parent.entries.set(entry.name, entry);
        parent.mtime = Date.now();
        parent.size += 1;

        this._fireSoon({ type: vscode.FileChangeType.Changed, uri: dirname }, { uri, type: vscode.FileChangeType.Deleted });
        
    }

    // --- lookup

    _lookup(uri, silent){
        const parts = uri.path.split('/');
        let entry = this.root;
        for (const part of parts) {
            if (!part) {
                continue;
            }
            let child;
            if (entry instanceof Directory) {
                child = entry.entries.get(part);
            }
            if (!child) {
                if (!silent) {
                    throw vscode.FileSystemError.FileNotFound(uri);
                } else {
                    return undefined;
                }
            }
            entry = child;
        }
        return entry;
    }

    _lookupAsDirectory(uri, silent) {
        const entry = this._lookup(uri, silent);
        if (entry instanceof Directory) {
            return entry;
        }
        throw vscode.FileSystemError.FileNotADirectory(uri);
    }

    _lookupAsFile(uri, silent) {
        const entry = this._lookup(uri, silent);
        if (entry instanceof File) {
            return entry;
        }
        throw vscode.FileSystemError.FileIsADirectory(uri);
    }

    _lookupParentDirectory(uri) {
        console.log(uri)
        const dirname = uri.with({ path: path.posix.dirname(uri.path) });
        return this._lookupAsDirectory(dirname, false);
    }

    // --- manage file events
    watch(_resource) {
        // ignore, fires for all changes...
        return new vscode.Disposable(() => { });
    }

    _fireSoon(...events) {
        this._bufferedEvents.push(...events);

        if (this._fireSoonHandle) {
            clearTimeout(this._fireSoonHandle);
        }

        this._fireSoonHandle = setTimeout(() => {
            this._emitter.fire(this._bufferedEvents);
            this._bufferedEvents.length = 0;
        }, 5);
    }
}

module.exports = {
    JHubFS,
    File,
    Directory
}